File: backend-ex/models/productModel.ts
import { Model, DataTypes, Optional } from 'sequelize';
import sequelize from '../config/database';
import { ProductImage } from './productImage';
import { ProductSize } from './productSize';

interface ProductAttributes {
  id: number;
  name: string;
  description: string;
  price: number;
  category: string;
  subCategory: string;
  bestseller: boolean;
}
interface ProductCreationAttributes extends Optional<ProductAttributes, 'id'> {}

export class Product extends Model<ProductAttributes, ProductCreationAttributes> implements ProductAttributes {
  public id!: number;
  public name!: string;
  public description!: string;
  public price!: number;
  public category!: string;
  public subCategory!: string;
  public bestseller!: boolean;

  // associations
  public images?: ProductImage[];
  public sizes?: ProductSize[];
}

Product.init(
  {
    id: { type: DataTypes.INTEGER.UNSIGNED, autoIncrement: true, primaryKey: true },
    name: { type: DataTypes.STRING, allowNull: false },
    description: { type: DataTypes.TEXT, allowNull: true },
    price: { type: DataTypes.DECIMAL(10, 2), allowNull: false },
    category: { type: DataTypes.STRING, allowNull: true },
    subCategory: { type: DataTypes.STRING, allowNull: true },
    bestseller: { type: DataTypes.BOOLEAN, defaultValue: false },
  },
  { sequelize, 
    tableName: 'products',
    timestamps: true,}
);

Product.hasMany(ProductImage, { foreignKey: 'productId', as: 'images' });
Product.hasMany(ProductSize, { foreignKey: 'productId', as: 'sizes' });

File: backend-go/models/product_model.go
package models

// Product represents the `products` table.
//
// ┌──────────────┐          ┌────────────────┐
// │  products    │ 1    ✱  │ product_images │
// └──────────────┘          └────────────────┘
//
// ┌──────────────┐          ┌────────────────┐
// │  products    │ 1    ✱  │ product_sizes  │
// └──────────────┘          └────────────────┘
type Product struct {
	ID          uint           `gorm:"primaryKey;autoIncrement;type:int" json:"id"`
	Name        string         `gorm:"type:varchar(255);not null" json:"name"`
	Description string         `gorm:"type:text" json:"description"`
	Price       float64        `gorm:"type:decimal(10,2);not null" json:"price"`
	Category    string         `gorm:"type:varchar(255)" json:"category"`
	SubCategory string         `gorm:"column:subCategory;type:varchar(255)" json:"subCategory"`
	Bestseller  bool           `gorm:"default:false" json:"bestseller"`
	Images      []ProductImage `gorm:"foreignKey:ProductID;constraint:OnDelete:CASCADE;" json:"images"`
	Sizes       []ProductSize  `gorm:"foreignKey:ProductID;constraint:OnDelete:CASCADE;" json:"sizes"`
}
